_source/_posts/2021-07-14-flask-scim-server.md

```diff
@@ -0,0 +1,519 @@
+---
+layout: blog_post
+title: "Okta, SCIM and Flask"
+author: cale-switzer
+by: advocate
+communities: [python]
+description: "Using a Flask SCIM server, learn more about the how & why behind how Okta implements the provisioning protocol."
+tags: [scim, python, flask]
+tweets:
+- ""
+- ""
+- ""
+image:
+type: awareness
+---
+
+At a very high level, the SCIM (System for Cross-Identity Management) protocol exists to provision and sync users and groups across various, independent systems. There is some room for nuance in the [SCIM spec](https://datatracker.ietf.org/doc/html/rfc7644). In this post we will take a look at some basic SCIM operations and what the flow looks like under the hood between Okta and a SCIM server built in Flask using PostgreSQL as our app database. Even if you're not a Python dev, hopefully you can take away some of the concepts in this guide for your own SCIM implementation.
+
+### What you'll need
+
+- [virtualenv](https://docs.python.org/3/library/venv.html)
+- [PostgreSQL](https://www.postgresql.org/)
+- [ngrok](https://ngrok.com/)
+
+### Setting up the Flask server and Postgres DB 
+{% img blog/flask-scim-server/flask_postgres.png alt:"PostgreSQL and Flask Logos Image" width:"500" %}
+
+1.  Clone the repo found [here](https://github.com/Tennyx/flask-scim-server), open terminal and `cd` into the project root.
+2.  Create a new virtualenv in the root folder with `virtualenv env`
+3.  Run the virtual environment with `source env/bin/activate`
+4.  Install necessary Python packages with ```pip install -r requirements.txt```
+5.  Create a new Postgres database named `scim`. Enter the `psql` shell by opening a new terminal tab and typing `psql postgres`. Create the DB with `CREATE DATABASE scim;` (Run `\l` to double-check the database has been created)
+6.  Go back to the terminal tab that is in the flask app root. Run the following commands to create migrations and tables in the ```scim``` database:
+    - `python manage.py db init`
```

Comment:
FYI, when I ran this command, I got this error:

```
$ python manage.py db init
/Users/joel/Projects/flask-scim-server/venv/lib/python3.8/site-packages/flask_sqlalchemy/__init__.py:833: FSADeprecationWarning: SQLALCHEMY_TRACK_MODIFICATIONS adds significant overhead and will be disabled by default in the future.  Set it to True or False to suppress this warning.
  warnings.warn(FSADeprecationWarning(
  Creating directory /Users/joel/Projects/flask-scim-server/migrations ...  done
  Creating directory /Users/joel/Projects/flask-scim-server/migrations/versions ...  done
  Generating /Users/joel/Projects/flask-scim-server/migrations/script.py.mako ...  done
  Generating /Users/joel/Projects/flask-scim-server/migrations/env.py ...  done
  Generating /Users/joel/Projects/flask-scim-server/migrations/README ...  done
  Generating /Users/joel/Projects/flask-scim-server/migrations/alembic.ini ...  done
  Please edit configuration/connection/logging settings in '/Users/joel/Projects/flask-scim-server/migrations/alembic.ini' before proceeding.
```

Reply:
Thanks for catching! I fixed this based off the following solution: https://stackoverflow.com/a/33790196/8025504